LibraryInstrumentationGuidance.md

```diff
@@ -1,53 +1,115 @@
 
-[Activity][ActivityGuide] is a class that allows storing and accessing diagnostics context and consuming it with logging system.
+[Activity][ActivityGuide] is a class that allows storing and accessing diagnostics context and consuming it with logging system. Activity flows with async calls and anything can request Activity.Current and correlate the trace/log with it.
 
 Both Diagnostic Source and Activity have been used to instrument [System.Net.Http][SystemNetHttp] and [Microsoft.AspNetCore.Hosting][MicrosoftAspNetCoreHosting].
-
 More recently two new libraries were instrumented and that work was the basis for this guidance. These libraries are client SDKs for [Azure Event Hubs][MicrosoftEventHubs] and [Azure Service Bus][MicrosoftServiceBus], both of which support high throughput scenarios.
 
-[This document][DiagnosticSourceActivityHowto] goes into more details on how to efficiently use Diagnostic Source.
-
 ## What should be instrumented
 
 The goal of instrumentation is to give the users the visibility to how particular operations are being performed inside the library. This information can be later used to diagnose performance or issues. It is up to the library authors to identify operations that the library performs and are worthy of monitoring. These operations can match the exposed API but also cover more specific internal logic (like outgoing service calls, retries, locking, cache utilization, awaiting system events, etc.). This way the users can get a good understanding of what's going on under the hood when they need it. 
 
-## Instrumentation 
+## Bare Minimum Instrumentation 
 
-In the simplest case, the operation that is being monitored has to be wrapped by an activity. However in order to minimize performance impact the activity should be only created if there is any listener waiting for it.
+In the simplest case, the operation that is being monitored has to be wrapped by an activity. 
 
 ```csharp
+    // create DiagnosticListener with unique name
     static DiagnosticListener source = new DiagnosticListener("Example.MyLibrary");
 
     Activity activity = null;
-    // create activity only when requested
-    if (source.IsEnabled() && source.IsEnabled("Example.MyLibrary.MyOperation"))
+    
+    // check if listener is interested in this operation
+    if (source.IsEnabled() && source.IsEnabled("MyOperation"))
     {
-        activity = new Activity("Example.MyLibrary.MyOperation");
-        source.StartActivity(activity, new { Input = input });
+        activity = new Activity("MyOperation");
+        activity.Start();
     }
 
     object output = null;
     try
     {
         // perform the actual operation
-        output = RunOperation(input);  
+        output = RunOperation(input);
+    }
+    catch (Exception ex)
+    {
+        activity?.AddTag("error", true);
```

Comment:
Tags values are strings so this has to be 

activity?.AddTag("error", "true");

Reply:
We should probably add special helpers like 'AddErrorTag' for canonical tags as we find they are very generic.   That makes the code simpler and allows Activity to encode it more efficiently in the future.

Reply:
having 'status' on the Activity would be great once we agree on the contract!

TL;DR;
So, I would  recommend setting a status code and optionally sending exception event.
 
Status should still provide "OK" or "something else" (like [OpenCensus](https://opencensus.io/trace.html) does). Having [strict enum-kind of mapping](https://github.com/census-instrumentation/opencensus-specs/blob/33b9a1af01c0cfb36036471fa9c464c7e7c6494d/trace/HTTP.md#mapping-from-http-status-codes-to-trace-status-codes) could be an option too, but it seems hard to do this mapping for each exception in the library 

Since the result of operation is special and essential, I think we should have a `Status` property (and `SetStatus(string)` method) on the Activity.

-------
I put "error" as a placeholder and do not think we should use it. Boolean value as [OpenTracing](https://github.com/opentracing/specification/blob/master/semantic_conventions.md#span-tags-table) suggests here does not seem to be informative enough: HTTP, gRPC and perhaps other communication would rather provide a status code. [OpenCensus](https://opencensus.io/trace.html) suggests status code + details.

While it is useful to stick to the tags suggested by [OpenTracing](https://github.com/opentracing/specification/blob/master/semantic_conventions.md#span-tags-table), I don't really understand the guidance there as well.

@cwe1ss, could you please advise us, what is the OpenTracing recommendation for errors

I see suggestion to use 
```
event="error"
error.object=<error object instance>
```
where possible. What about generic (non http) status recommendation in the tags?

This recommendation translates to DiagnosticSource 'Exception' event rather than a tag on the Span/Activity.
After working with exception events for a while, I can see an issue with them:
We expected someone would be interested in exceptions only without Activities. It seems tracking ALL exceptions without context is not that useful - it could mean overwhelming amount of handled exceptions and we hear users are interested mostly in unhandled ones.
Because of the same assumptions they [complicate instrumentation](https://github.com/Azure/azure-service-bus-dotnet/blob/dev/src/Microsoft.Azure.ServiceBus/Core/MessageReceiver.cs#L383) a bit.

Reply:
#### Status code

OpenTracing currently only defines the "http.status_code" header and no non-http status codes. As there's no real protocol-independent standard (or none that I know of), I doubt that it makes sense to have a generic "status" tag/property. Instead, I think that libraries that have status codes should add their own tags - e.g. "grpc.status_code".

#### Errors
As far as I understand, the OpenTracing "error" tag is for "Exception"-like cases, yes. Most OpenTracing instrumentation I've seen adds the "Exception" details via a `span.Log( "event"="error", "error.object"=exception )` call (pseudo code). 
Of course, the question is, whether this a case of the "log & throw" anti-pattern. If there's a parent span/activity it might log the same exception and re-throw again. 

Not sure what the best solution is here. The instrumentation code could only set the "error" tag and not log/raise the exception. This means it would rely on some parent span/activity do log it but what if there isn't one? Also, it might make debugging harder as the exception details are associated with an activity/span that did not raise it. Imagine there's a retry-scenario where something fails twice and succeeds on the third attempt. Who's responsible for logging the exceptions?

There's also a difference between OpenTracing and Activity because with OpenTracing's "Log" concept, the exception always stays with the span whereas this is more decoupled in `Activity` where it currently can only be done via a separate DiagnosticSource event.